leader/leader.go

```diff
@@ -88,241 +87,282 @@ type CallbackFn func(key, prevValue, newValue string)
-			select {
-			case valuesC <- resp.Node.Value:
-			case <-l.closeC:
-				return
-			}
-		}
+// RemoveVoter stops the voting loop.
+func (l *Client) RemoveVoter(ctx context.Context, key, value string, term time.Duration) {
+	l.once.Do(func() {
+		l.startVoterLoop(key, value, term, false)
+	})
+	select {
+	case l.voterC <- false:
+	case <-ctx.Done():
+	case <-l.closeC:
+	}
+}
 
-		var sentAnything bool
-		for {
+// StepDown makes this participant to pause his attempts to re-elect itself thus giving up its leadership
+func (l *Client) StepDown(ctx context.Context) {
+	select {
+	case l.pauseC <- true:
+	case <-ctx.Done():
+	}
+}
 
-			if watcher == nil {
-				watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-			}
+// Close stops current operations and releases resources
+func (l *Client) Close() error {
+	// already closed
+	if !atomic.CompareAndSwapUint32(&l.closed, 0, 1) {
+		return nil
+	}
+	close(l.closeC)
+	l.wg.Wait()
+	return nil
+}
 
-			if watcher != nil {
-				resp, err = watcher.Next(ctx)
-				if err == nil {
-					if resp.Node.Value == "" {
-						continue
-					}
-					backoff.Reset()
-				}
-			}
+func (l *Client) startVoterLoop(key, value string, term time.Duration, enabled bool) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-l.closeC
+		cancel()
+	}()
+	go l.voterLoop(ctx, key, value, term, enabled)
+}
 
+func (l *Client) watchLoop(ctx context.Context, key string, valuesC chan string, logger logrus.FieldLogger) {
+	defer l.wg.Done()
+	boff := newBackoff()
+	// maxFailedSteps sets the limit on the number of failed attempts before the watch
+	// is forcibly reset
+	const maxFailedSteps = 10
+	var (
+		api     = client.NewKeysAPI(l.Client)
+		watcher client.Watcher
+	)
+	var err error
+	for {
+		select {
+		case <-time.After(boff.NextBackOff()):
+		case <-l.closeC:
+			return
+		}
+		if watcher == nil {
+			watcher, err = l.getWatchAtLatestIndex(ctx, api, key, valuesC, logger)
+			log.WithError(err).Info("Create watcher at latest index.")
 			if err != nil {
-				select {
-				case <-ticker.C:
-					steps += 1
-				}
-
-				if err == context.Canceled {
+				if IsContextError(err) {
 					return
-				} else if cerr, ok := err.(*client.ClusterError); ok {
-					if len(cerr.Errors) != 0 && cerr.Errors[0] == context.Canceled {
-						return
-					}
-					log.Debugf("unexpected cluster error: %v (%v)", err, cerr.Detail())
-					continue
-				} else if IsWatchExpired(err) {
-					log.Debug("watch expired, resetting watch index")
-					watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-					if err != nil {
-						continue
-					}
+				}
+				if IsWatchExpired(err) {
+					// The watcher has expired, reset it so it's recreated on the
+					// next loop cycle.
+					logger.Info("Watch has expired, resetting watch index.")
+					watcher = nil
+				} else if err := asClusterError(err); err != nil {
+					logger.Warnf("Cluster error: %v.", err)
```

Comment:
Shouldn't we reset the watcher or increase the number of attempts in this case?

Reply:
Not sure but I think we can remove the `cluster error` special case and just treat the rest as unknown and bump/reset the backoff.